plugins/resource_monitor_plugin/resource_monitor_plugin.cpp

```diff
@@ -0,0 +1,228 @@
+/**
+   It was reported from a customer that when file system which 
+   "data/blocks" belongs to is running out of space, the producer  
+   continued to produce blocks and update state but the blocks log was
+   "corrupted" in that it no longer contained all the irreversible blocks.
+   It was also observed that when file system which "data/state"
+   belons to is running out of space, nodeos will crash with SIGBUS as
+   the state file is unable to acquire new pages.
+
+   The solution is to have a dedicated plugin to monitor resource 
+   usages (file system space now, CPU, memory, and networking
+   bandwidth in the future).
+   The plugin uses a thread to periodically check space usage of file
+   systems of directories being monitored. If space used
+   is over a predefined threshold, a graceful shutdown is initiated.
+
+   The directories to be monitored are configured;
+   they can be any combination of data, data/blocks, data/state and
+   any directories like those for trace and state history.
+   Space threshold can also be configured.
+
+   Alternative solution to blocks directory space exhaustion
+   was considered: starting from block_log::append, 
+   make sure all exceptions was handled properly. But it was deemed to be
+   too tedious and error prone to review all possible excpetions, and
+   might likely break concensus. The plugin approach is more generic,
+   monitoring any file systems and without modifying existing code.
+
+**/
+
+#include <eosio/resource_monitor_plugin/resource_monitor_plugin.hpp>
+#include <eosio/chain/exceptions.hpp>
+
+#include <fc/exception/exception.hpp>
+#include <fc/log/logger_config.hpp> // set_os_thread_name
+
+#include <boost/filesystem.hpp> // C++17's filesystem not in our toolchain, use Boost's
+
+#include <thread>
+
+#include <sys/stat.h>
+
+namespace bfs = boost::filesystem;
+
+namespace eosio {
+   static appbase::abstract_plugin& _resource_monitor_plugin = app().register_plugin<resource_monitor_plugin>();
+
+class resource_monitor_plugin_impl {
+public:
+
+   void set_program_options(options_description&, options_description& cfg) {
+      cfg.add_options()
+         ( "resource-monitor-directory",  bpo::value<vector<bfs::path>>()->composing(),
+           "The directory whose file system to be monitored. This is a mandatory option. It can appear multiple times for different directories")
+         ( "resource-monitor-interval-seconds", bpo::value<uint32_t>()->default_value(2),
+           "Time in seconds between two consecutive checks of resource usage. Should be greater than 1 and less than 60 (one minute)" )
+         ( "resource-monitor-space-threshold", bpo::value<uint32_t>()->default_value(95),
```

Comment:
default is still 95 here, which I think will overwrite the 90 that is intended

Reply:
Thanks Matt! Sorry I should have used constants for those. Will change them.